Drivers: ADC: Add ADC Driver for Infineon PSC3 Series

jsbatch opened the pull request:
This change implements an ADC driver for the PSOC C3 family of devices using the High Performance Programmable Analog Subsystem (HPPASS).

- Updates device tree and KConfig files based on new ADC driver implementation.
- Updates modules/hal/infineon to include new support files.
- Adds HPPASS SAR ADC driver files used in PSOC C3 series of devices.
- Adds sample and test overlays to provide support for: adc_api

jsbatch committed:
dts: infineon: psc3m5 devicetree changes to support HPPASS ADC

 * Separates HPPASS and HPPASS SAR ADC in the device tree
 * Makes HPPASS SAR ADC a child of the HPPASS system to reflect hardware
   architecture.
 * Adds binding files for HPPASS SAR ADC driver.

Signed-off-by: John Batch <john.batch@infineon.com>

jsbatch committed:
drivers: adc: Infineon HPPASS SAR ADC Driver

Adds HPPASS SAR ADC driver and HPPASS Analog driver files to support
ADC conversion for the PSOC C3 family of MCUs.

Signed-off-by: John Batch <john.batch@infineon.com>

JarmouniA reviewed (commented):
Separate samples and tests changes into 2 commits.

jsbatch committed:
drivers: adc: Adding Infineon HPPASS ADC samples

Adding files to allow running the following samples for the
Infineon HPPASS SAR ADC driver on the PSOC C3 family of MCUs:
  * samples/drivers/adc_dt
  * samples/drivers/adc_sequence

Signed-off-by: John Batch <john.batch@infineon.com>

jsbatch committed:
drivers: adc: Adding Infineon HPPASS ADC tests

Adding files to allow running the following tests for the Infineon
HPPASS SAR ADC driver on the PSOC C3 family of MCUs:
  * tests/drivers/adc/adc_api
  * tests/drivers/adc/adc_error_case
Modifies the adc_error_cases test to use add a KConfig option
for the valid resolution.

Signed-off-by: John Batch <john.batch@infineon.com>

jsbatch commented on tests/drivers/adc/adc_error_cases/Kconfig:
```diff
@@ -0,0 +1,10 @@
+
+config ADC_ERROR_CASES_RESOLUTION
+	int "ADC valid resolution"
+	default 10
```
This has been changed.

sreeramIfx commented on drivers/adc/adc_ifx_hppass_sar.c:
```diff
@@ -0,0 +1,765 @@
+ * channels are initially disabled.  Channels will be enabled and configured in the channel setup
+ * function.
+ */
+static void _ifx_init_channel_cfg(struct ifx_hppass_sar_adc_data *data)
```
```suggestion
static void ifx_init_channel_cfg(struct ifx_hppass_sar_adc_data *data)
```

jsbatch replied:
Fixed

jsbatch commented on drivers/adc/adc_ifx_hppass_sar.c:
```diff
@@ -0,0 +1,765 @@
+ * be configured later in the channel setup function.  Groups are configured when a conversion is
+ * started.
+ */
+static void _ifx_init_pdl_struct(struct ifx_hppass_sar_adc_data *data,
```
Fixed

JarmouniA commented on dts/arm/infineon/cat1b/psc3/psc3_s.dtsi:
```diff
@@ -128,12 +128,20 @@
-			interrupts = <109 4>;
-			status = "disabled";
-			#io-channel-cells = <1>;
+		hppass_analog0: hppass-analog@52b00000 {
```
```suggestion
		hppass_analog0: analog@52b00000 {
```

jsbatch replied:
I've made these updates.
